proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,128 @@
+
+## 1. Motivation
+
+Since its inception, the Haskell programming language and GHC, its principal toolchain/development platform, has been open to innovation in programming language design through Haskell language extensions. The creation of those extensions has been well managed, giving rise to a programming language that is both coherent and still evolving and vital after more than three decades of astonishing development.
+
+However, once an extension clears the review process and gets implemented in the compiler, there is currently no framework for understanding or managing its evolution. The strength of the development model is its openness, allowing experimental features to be tried out in the community. Features that the community discovers to be problematic in practice can be not enabled and the community can, over time, reach a consensus about the value of these extensions.
+
+The weakness of this model is that, over time, the accumulation of programming extensions in various states of use and disuse can create confusion, especially for developers being introduced to Haskell. There is not one Haskell, but many, and each individual or team must necessarily choose their own. Not everyone has yet had the opportunity to accumulate the experience and knowledge that will allow them to quickly determine whether or not to use a given language extension, leading to various points of stress. 
+
+The authors of this proposal believe that a clear lifecycle for these programming language extensions would make it easier for newcomers, those responsible for onboarding new Haskell developers, and the community more broadly, to make sense of the Haskell language extensions. This framework would make it clear which extensions are experimental and prone to rapid evolution, which of them are mature and thus relatively stable and suitable for serious use, and those that are generally not considered suitable for new development. Extending GHC's warnings to account for these categories allows projects to find out when features they use have been deprecated in a new compiler version as well as to specify general policies about language stability that can be checked mechanically.
+
+## 2. Proposed Change Specification
+
+### 2.1. Categories and Lifecycle
+Language extensions will be classified into the following categories
+ * `Experimental` extensions are undergoing active development. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental. At the time of writing, `LinearTypes` and `OverloadedRecordUpdate` seem to be in this category.
```

Comment:
`Experimental` covers two possibly non-overlapping categories: newer extensions and extensions that are likely to change. I definitely want to be warned about the latter, but not the former. Perhaps another category is in order: `Nascent`

Reply:
> Experimental covers two possibly non-overlapping categories: newer extensions and extensions that are likely to change. I definitely want to be warned about the latter, but not the former. Perhaps another category is in order: Nascent

Let's not have too many categories!  It would just get confusing and harder to decide which category to put something into.  Newer extensions are, because they are new, quite likely to change.  Less is more :-)

Perhaps it would help to frame the categories in actionable form:

* `Stable` (which I agree is better than `Mature`): you can rely on this; it's unlikely to change.
* `Experimental`: quite likely to change.
* `Deprecated`: try to avoid using this at all, and actively remove it from code you maintain.
* `Legacy`: avoid for new code.

Reply:
> > Experimental covers two possibly non-overlapping categories: newer extensions and extensions that are likely to change. I definitely want to be warned about the latter, but not the former. Perhaps another category is in order: Nascent
> 
> Let's not have too many categories! It would just get confusing and harder to decide which category to put something into. Newer extensions are, because they are new, quite likely to change. Less is more :-)

I agree with this completely. And if we learn more after living with this for a while, we can make adjustments!

> 
> Perhaps it would help to frame the categories in actionable form:
> 
> * `Stable` (which I agree is better than `Mature`): you can rely on this; it's unlikely to change.
`Stable` was specifically avoided based on prior feedback around language extension classification. That said, I personally agree that `Stable` is a better name, and would certainly endorse a change there.

> * `Deprecated`: try to avoid using this at all, and actively remove it from code you maintain.
And per the proposal, there is some statement about how long it will continue to exist.

> * `Legacy`: avoid for new code.
But it will continue to exist indefinitely.

Reply:
> `Stable` was specifically avoided based on prior feedback around language extension classification.

I'm curious to know what the arguments are against `Stable`. From a brief skim of the discussion there seem to be quite a few people favouring it over `Mature`. And it seems to more clearly communicate "this isn't expected to change", whereas `Mature` implies "this has been around a while" without necessarily indicating that it is in a good state to use (at least to me). For example, I would say `FunctionalDependencies` is a very "mature" extension but would hesitate to declare it "stable" given the number of dark corners it exposes.

Reply:
Isn't your last example precisely an argument for mature (implying sensible, reliable, not childish)? Stable only indicates ”unchanging”, but has otherwise no connotations of quality. All the legacy extensions are probably more stable than the mature ones…


To me, mature means “ready for production use” which _implies_ stable (or at least only backward compatible changes maybe), but says more actually.

But in the end it's just a name…

Reply:
> > `Stable` was specifically avoided based on prior feedback around language extension classification.
> 
> I'm curious to know what the arguments are against `Stable`. From a brief skim of the discussion there seem to be quite a few people favouring it over `Mature`. And it seems to more clearly communicate "this isn't expected to change", whereas `Mature` implies "this has been around a while" without necessarily indicating that it is in a good state to use (at least to me). For example, I would say `FunctionalDependencies` is a very "mature" extension but would hesitate to declare it "stable" given the number of dark corners it exposes.

I _believe_ most of the arguments against were around how to precisely define what `Stable` means. 

> 
> To me, mature means “ready for production use” which _implies_ stable (or at least only backward compatible changes maybe), but says more actually.
> 
> But in the end it's just a name…

This, along with some notion that stable was not a good name, is exactly how mature came to be the name. To me whichever way makes the understanding the easiest to the most people is the way to go, which feels like `Stable` at the moment.

Reply:
It's interesting how the same names can have different connotations to different people. :grin: I'm happy whichever way we end up painting this particular bikeshed.

Reply:
After consulting with my coauthors, I updated the proposal to call it `Unstable`. Thanks!